File: fabric/src/main/java/dev/kineticcat/hexportation/fabric/transfer/TransactionImpl.java
package dev.kineticcat.hexportation.fabric.transfer;

import dev.kineticcat.hexportation.api.transfer.Transaction;

/**
 * Fabric implementation using singleton injection pattern.
 */
public class TransactionImpl extends Transaction.Implementation {
    
    public static final TransactionImpl INSTANCE = new TransactionImpl();
    
    @Override
    public Transaction openOuter() {
        return new FabricTransactionWrapper(net.fabricmc.fabric.api.transfer.v1.transaction.Transaction.openOuter());
    }
    
    /**
     * Wrapper class that extends our Transaction abstract class using Fabric's Transaction.
     */
    public static class FabricTransactionWrapper extends Transaction {
        
        private final net.fabricmc.fabric.api.transfer.v1.transaction.Transaction fabricTransaction;
        
        public FabricTransactionWrapper(net.fabricmc.fabric.api.transfer.v1.transaction.Transaction fabricTransaction) {
            this.fabricTransaction = fabricTransaction;
        }
        
        @Override
        public Transaction openNested() {
            return new FabricTransactionWrapper(fabricTransaction.openNested());
        }
        
        @Override
        public void commit() {
            fabricTransaction.commit();
        }
        
        @Override
        public void abort() {
            fabricTransaction.abort();
        }
        
        @Override
        public void close() {
            fabricTransaction.close();
        }
        
        /**
         * Get the underlying Fabric transaction for platform-specific operations.
         */
        public net.fabricmc.fabric.api.transfer.v1.transaction.Transaction getFabricTransaction() {
            return fabricTransaction;
        }
    }
}

File: fabric/src/main/java/dev/kineticcat/hexportation/fabric/transfer/TransactionImpl.kt
package dev.kineticcat.hexportation.fabric.transfer

import dev.kineticcat.hexportation.api.transfer.Transaction
import net.fabricmc.fabric.api.transfer.v1.transaction.Transaction as FabricTransaction

/**
 * Fabric implementation using singleton injection pattern.
 */
class TransactionImpl : Transaction.Implementation() {
    
    companion object {
        @JvmField
        val INSTANCE = TransactionImpl()
    }
    
    override fun openOuter(): Transaction = 
        FabricTransactionWrapper(FabricTransaction.openOuter())
    
    /**
     * Wrapper class that extends our Transaction abstract class using Fabric's Transaction.
     */
    class FabricTransactionWrapper(
        private val fabricTransaction: FabricTransaction
    ) : Transaction() {
        
        override fun openNested(): Transaction = 
            FabricTransactionWrapper(fabricTransaction.openNested())
        
        override fun commit() = fabricTransaction.commit()
        
        override fun abort() = fabricTransaction.abort()
        
        override fun close() = fabricTransaction.close()
        
        /**
         * Get the underlying Fabric transaction for platform-specific operations.
         */
        fun getFabricTransaction() = fabricTransaction
    }
}